packages/firestore/src/core/component_provider.ts

```diff
@@ -0,0 +1,299 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ClientId,
+  MemorySharedClientState,
+  SharedClientState,
+  WebStorageSharedClientState
+} from '../local/shared_client_state';
+import { LocalStore, MultiTabLocalStore } from '../local/local_store';
+import { MultiTabSyncEngine, SyncEngine } from './sync_engine';
+import { RemoteStore } from '../remote/remote_store';
+import { EventManager } from './event_manager';
+import { AsyncQueue } from '../util/async_queue';
+import { DatabaseInfo } from './database_info';
+import { Platform } from '../platform/platform';
+import { Datastore } from '../remote/datastore';
+import { User } from '../auth/user';
+import { PersistenceSettings } from './firestore_client';
+import { assert } from '../util/assert';
+import { GarbageCollectionScheduler, Persistence } from '../local/persistence';
+import { Code, FirestoreError } from '../util/error';
+import { OnlineStateSource } from './types';
+import { LruParams, LruScheduler } from '../local/lru_garbage_collector';
+import { IndexFreeQueryEngine } from '../local/index_free_query_engine';
+import { IndexedDbPersistence } from '../local/indexeddb_persistence';
+import {
+  MemoryEagerDelegate,
+  MemoryPersistence,
+  MemoryReferenceDelegate
+} from '../local/memory_persistence';
+
+/**
+ * Initializes and wires up all core components for Firestore. Consumers have to
+ * invoke initialize() once before accessing any of the individual components.
+ */
+export abstract class ComponentProvider {
```

Comment:
Do you think this would end up as less code if this were just an interface that exposed the properties?

Reply:
I changed this into an abstract class to share the asserts. I don't feel strongly.